src/main/kotlin/org/jetbrains/research/libsl/nodes/references/TypeReference.kt

```diff
@@ -3,55 +3,152 @@ package org.jetbrains.research.libsl.nodes.references
 import org.jetbrains.research.libsl.context.LslContextBase
 import org.jetbrains.research.libsl.type.*
 
-abstract class TypeReference(
-    open val name: String,
-    open val isPointer: Boolean = false,
-    open var typeBound: GenericTypeBound = GenericTypeBound.EMPTY,
-    open val genericReferences: MutableList<TypeReference> = mutableListOf(),
+interface TypeReference : LslReference<Type, TypeReference> {
     override val context: LslContextBase
-) : LslReference<Type, TypeReference> {
+}
+
+data class UnionExpressionTypeReference(
+    val left: TypeReference,
+    val right: TypeReference,
+    override val context: LslContextBase
+) : TypeReference {
+    override fun resolve(): Type? {
+        this.left.resolve()
+        this.right.resolve()
+        // We don't have type for UnionExpression
+        return null
+    }
+
+    override fun isReferenceMatchWithNode(node: Type): Boolean {
+        // we don't have Type for UnionExpressionTypeReference
+        return false
+    }
```

Comment:
May be something like this instead? (I don't have full context in mind)
```suggestion
    override fun isReferenceMatchWithNode(node: Type): Boolean {
        return left.isReferenceMatchWithNode(node) || right.isReferenceMatchWithNode(node)
    }
```

Reply:
Fixed;